# **1. Import Library**

Pada tahap ini, Anda perlu mengimpor beberapa pustaka (library) Python yang dibutuhkan untuk analisis data dan pembangunan model machine learning.

In [25]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split
from sklearn import tree
from sklearn import metrics
from sklearn.pipeline import Pipeline
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

# **2. Memuat Dataset dari Hasil Clustering**

Memuat dataset hasil clustering dari file CSV ke dalam variabel DataFrame.

In [26]:
import pandas as pd
df = pd.read_csv("hasil_clustering.csv")
df.head()

BALANCE  BALANCE_FREQUENCY  PURCHASES  ONEOFF_PURCHASES  \
0    40.900749           0.818182      95.40              0.00   
1  3202.467416           0.909091       0.00              0.00   
2  2495.148862           1.000000     773.17            773.17   
3  1666.670542           0.636364    1499.00           1499.00   
4   817.714335           1.000000      16.00             16.00   

   INSTALLMENTS_PURCHASES  CASH_ADVANCE  PURCHASES_FREQUENCY  \
0                    95.4      0.000000             0.166667   
1                     0.0   6442.945483             0.000000   
2                     0.0      0.000000             1.000000   
3                     0.0    205.788017             0.083333   
4                     0.0      0.000000             0.083333   

   ONEOFF_PURCHASES_FREQUENCY  PURCHASES_INSTALLMENTS_FREQUENCY  \
0                    0.000000                          0.083333   
1                    0.000000                          0.000000   
2                    1.000000                          0.000000   
3                    0.083333                          0.000000   
4                    0.083333                          0.000000   

   CASH_ADVANCE_FREQUENCY  CASH_ADVANCE_TRX  PURCHASES_TRX  CREDIT_LIMIT  \
0                0.000000                 0              2        1000.0   
1                0.250000                 4              0        7000.0   
2                0.000000                 0             12        7500.0   
3                0.083333                 1              1        7500.0   
4                0.000000                 0              1        1200.0   

      PAYMENTS  MINIMUM_PAYMENTS  PRC_FULL_PAYMENT  TENURE  Cluster  
0   201.802084        139.509787          0.000000      12        1  
1  4103.032597       1072.340217          0.222222      12        1  
2   622.066742        627.284787          0.000000      12        1  
3     0.000000        312.343947          0.000000      12        1  
4   678.334763        244.791237          0.000000      12        1

# **3. Data Splitting**

Tahap Data Splitting bertujuan untuk memisahkan dataset menjadi dua bagian: data latih (training set) dan data uji (test set).

In [27]:
X = df.drop(['Cluster'], axis=1)
y = df[['Cluster']]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=3)

In [28]:
X_train

BALANCE  BALANCE_FREQUENCY  PURCHASES  ONEOFF_PURCHASES  \
5734  1079.097023                1.0     247.44              0.00   
6631  2590.469749                1.0     459.12            422.00   
8659   945.802599                1.0     540.00              0.00   
8503  3074.490820                1.0       0.00              0.00   
7433  1325.679991                1.0    1084.87            687.57   
...           ...                ...        ...               ...   
7161  1653.400854                1.0     100.00            100.00   
2707   276.446975                1.0    1429.39            100.00   
6400  1434.793889                1.0     467.88            467.88   
1688  4530.639094                1.0       0.00              0.00   
5994   788.425212                1.0      94.38             32.08   

      INSTALLMENTS_PURCHASES  CASH_ADVANCE  PURCHASES_FREQUENCY  \
5734                  247.44      0.000000             1.000000   
6631                   37.12      0.000000             0.333333   
8659                  540.00    912.094948             1.000000   
8503                    0.00   3382.232089             0.000000   
7433                  397.30   3330.913556             0.833333   
...                      ...           ...                  ...   
7161                    0.00   2031.677435             0.090909   
2707                 1329.39      0.000000             1.000000   
6400                    0.00      0.000000             1.000000   
1688                    0.00   3162.719581             0.000000   
5994                   62.30   1461.675329             0.166667   

      ONEOFF_PURCHASES_FREQUENCY  PURCHASES_INSTALLMENTS_FREQUENCY  \
5734                    0.000000                          1.000000   
6631                    0.333333                          0.083333   
8659                    0.000000                          0.875000   
8503                    0.000000                          0.000000   
7433                    0.333333                          0.833333   
...                          ...                               ...   
7161                    0.090909                          0.000000   
2707                    0.083333                          1.000000   
6400                    1.000000                          0.000000   
1688                    0.000000                          0.000000   
5994                    0.083333                          0.166667   

      CASH_ADVANCE_FREQUENCY  CASH_ADVANCE_TRX  PURCHASES_TRX  CREDIT_LIMIT  \
5734                0.000000                 0             12        1200.0   
6631                0.000000                 0             11        2900.0   
8659                0.125000                 4             15        1000.0   
8503                0.166667                 2              0        4000.0   
7433                0.416667                 9             16        3000.0   
...                      ...               ...            ...           ...   
7161                0.363636                13              1        2000.0   
2707                0.000000                 0             21        2800.0   
6400                0.000000                 0             12        4300.0   
1688                0.250000                 5              0        8000.0   
5994                0.500000                10              3        2500.0   

         PAYMENTS  MINIMUM_PAYMENTS  PRC_FULL_PAYMENT  TENURE  
5734   503.445403        302.481716          0.000000      12  
6631   584.707120       1369.991379          0.000000      12  
8659   365.497356       1236.864840          0.000000       8  
8503  3249.404994        695.004032          0.000000      12  
7433  4135.414563        405.944341          0.090909      12  
...           ...               ...               ...     ...  
7161   575.557594        432.921030          0.000000      11  
2707  1296.041874        193.607553          0.100000      12  
6400  2538

In [29]:
X_test

BALANCE  BALANCE_FREQUENCY  PURCHASES  ONEOFF_PURCHASES  \
5535    76.673877           0.777778     410.90              0.00   
7828   645.406005           1.000000       0.00              0.00   
1680  4977.237859           1.000000      61.99             61.99   
4518   380.517893           1.000000    2590.27           2102.46   
6131    41.011184           1.000000    1314.75            125.19   
...           ...                ...        ...               ...   
265   8097.334733           1.000000    3999.92            200.00   
5034    13.221335           0.857143     108.84              0.00   
1278  1770.072581           1.000000       0.00              0.00   
285   4246.430225           1.000000       0.00              0.00   
1120  4688.103434           1.000000       0.00              0.00   

      INSTALLMENTS_PURCHASES  CASH_ADVANCE  PURCHASES_FREQUENCY  \
5535                  410.90      0.000000             0.666667   
7828                    0.00    929.220085             0.000000   
1680                    0.00   7448.329320             0.083333   
4518                  487.81      0.000000             0.916667   
6131                 1189.56      0.000000             1.000000   
...                      ...           ...                  ...   
265                  3799.92      0.000000             1.000000   
5034                  108.84      0.000000             0.857143   
1278                    0.00    781.616559             0.000000   
285                     0.00   9270.408108             0.000000   
1120                    0.00    357.129377             0.000000   

      ONEOFF_PURCHASES_FREQUENCY  PURCHASES_INSTALLMENTS_FREQUENCY  \
5535                    0.000000                          0.555556   
7828                    0.000000                          0.000000   
1680                    0.083333                          0.000000   
4518                    0.916667                          0.416667   
6131                    0.500000                          1.000000   
...                          ...                               ...   
265                     0.083333                          1.000000   
5034                    0.000000                          0.714286   
1278                    0.000000                          0.000000   
285                     0.000000                          0.000000   
1120                    0.000000                          0.000000   

      CASH_ADVANCE_FREQUENCY  CASH_ADVANCE_TRX  PURCHASES_TRX  CREDIT_LIMIT  \
5535                0.000000                 0              6        1500.0   
7828                0.333333                 4              0        1000.0   
1680                0.500000                22              1        6500.0   
4518                0.000000                 0             49        2500.0   
6131                0.000000                 0             70        6000.0   
...                      ...               ...            ...           ...   
265                 0.000000                 0             14        7500.0   
5034                0.000000                 0              9        1500.0   
1278                0.083333                 1              0        2500.0   
285                 0.666667                16              0        7000.0   
1120                0.166667                 2              0        5000.0   

         PAYMENTS  MINIMUM_PAYMENTS  PRC_FULL_PAYMENT  TENURE  
5535   449.090030        100.105897          0.000000       9  
7828   193.013564        310.454708          0.000000      12  
1680  6850.235626       1587.042735          0.000000      12  
4518  1825.405687        173.164117          0.090909      12  
6131  1303.104693        135.761304          0.700000      12  
...           ...               ...               ...     ...  
265   5624.074339       9952.672511          0.000000      12  
5034    69.973565         85.417572          0.000000       7  
1278  1390

# **4. Membangun Model Klasifikasi**


## **a. Membangun Model Klasifikasi**

Setelah memilih algoritma klasifikasi yang sesuai, langkah selanjutnya adalah melatih model menggunakan data latih.

Berikut adalah rekomendasi tahapannya.
1. Pilih algoritma klasifikasi yang sesuai, seperti Logistic Regression, Decision Tree, Random Forest, atau K-Nearest Neighbors (KNN).
2. Latih model menggunakan data latih.

In [30]:
# Simpan model dalam dictionary untuk mempermudah pengelolaan
models = {
    "Decision Tree": DecisionTreeClassifier(criterion="entropy"),  # Model Decision Tree
    "Random Forest": RandomForestClassifier(random_state=42)       # Model Random Forest
}

In [31]:
#Simpan hasil prediksi dari setiap model
predictions = {}

In [32]:
#Latih setiap model dan lakukan prediksi
for name, model in models.items():
    model.fit(X_train, y_train)  # Melatih model menggunakan data latih
    predictions[name] = model.predict(X_test)  # Simpan hasil prediksi

/usr/local/lib/python3.10/dist-packages/sklearn/base.py:1389: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return fit_method(estimator, *args, **kwargs)


Tulis narasi atau penjelasan algoritma yang Anda gunakan.

Dalam membangun model ini, saya menggunakan dua algoritma populer dari Machine Learning untuk klasifikasi, yaitu Decision Tree dan Random Forest.

## **b. Evaluasi Model Klasifikasi**

Berikut adalah **rekomendasi** tahapannya.
1. Lakukan prediksi menggunakan data uji.
2. Hitung metrik evaluasi seperti Accuracy dan F1-Score (Opsional: Precision dan Recall).
3. Buat confusion matrix untuk melihat detail prediksi benar dan salah.

In [33]:
#Evaluasi performa setiap model
for name, y_pred in predictions.items():
    print(f"\nEvaluasi Model: {name}")  # Tampilkan nama model
    print("Confusion Matrix:")
    print(confusion_matrix(y_test, y_pred))  # Tampilkan Confusion Matrix
    print("Classification Report:")
    print(classification_report(y_test, y_pred))  # Tampilkan Classification Report


Evaluasi Model: Decision Tree
Confusion Matrix:
[[ 111    8]
 [   9 2557]]
Classification Report:
              precision    recall  f1-score   support

           0       0.93      0.93      0.93       119
           1       1.00      1.00      1.00      2566

    accuracy                           0.99      2685
   macro avg       0.96      0.96      0.96      2685
weighted avg       0.99      0.99      0.99      2685


Evaluasi Model: Random Forest
Confusion Matrix:
[[ 109   10]
 [   4 2562]]
Classification Report:
              precision    recall  f1-score   support

           0       0.96      0.92      0.94       119
           1       1.00      1.00      1.00      2566

    accuracy                           0.99      2685
   macro avg       0.98      0.96      0.97      2685
weighted avg       0.99      0.99      0.99      2685



Tulis hasil evaluasi algoritma yang digunakan, jika Anda menggunakan 2 algoritma, maka bandingkan hasilnya.

- Random Forest menunjukkan performa yang sedikit lebih baik dibandingkan Decision Tree terutama pada F1-score dan pengurangan kesalahan prediksi.

- Random Forest lebih unggul karena menggunakan ensemble (gabungan banyak pohon), sehingga mampu memberikan hasil yang lebih stabil dan akurat dibandingkan Decision Tree yang hanya bergantung pada satu pohon keputusan.

- Namun, jika waktu pelatihan dan prediksi menjadi kendala, Decision Tree dapat menjadi pilihan yang lebih cepat dan sederhana.

## **c. Tuning Model Klasifikasi (Optional)**

Gunakan GridSearchCV, RandomizedSearchCV, atau metode lainnya untuk mencari kombinasi hyperparameter terbaik

In [34]:
#Type your code here

## **d. Evaluasi Model Klasifikasi setelah Tuning (Optional)**

Berikut adalah rekomendasi tahapannya.
1. Gunakan model dengan hyperparameter terbaik.
2. Hitung ulang metrik evaluasi untuk melihat apakah ada peningkatan performa.

In [35]:
#Type your code here

## **e. Analisis Hasil Evaluasi Model Klasifikasi**

Berikut adalah **rekomendasi** tahapannya.
1. Bandingkan hasil evaluasi sebelum dan setelah tuning (jika dilakukan).
2. Identifikasi kelemahan model, seperti:
  - Precision atau Recall rendah untuk kelas tertentu.
  - Apakah model mengalami overfitting atau underfitting?
3. Berikan rekomendasi tindakan lanjutan, seperti mengumpulkan data tambahan atau mencoba algoritma lain jika hasil belum memuaskan.

Kelas Minoritas (Kelas 0):
Pada kedua model, metrik untuk kelas 0 (minoritas) menunjukkan kelemahan:
Random Forest: Recall lebih rendah (92%) dibandingkan Precision (96%).
Decision Tree: Precision dan Recall untuk kelas 0 lebih rendah (93%) dibandingkan kelas 1.
Hal ini mengindikasikan bahwa kedua model mungkin mengalami bias terhadap kelas mayoritas (kelas 1).

Overfitting atau Underfitting:
Kedua model menunjukkan performa sangat baik di semua metrik. Namun, dengan accuracy mendekati 99%, ada potensi overfitting, terutama jika hasil pada dataset uji jauh lebih buruk dibandingkan dataset latih.